api/wallet_test.go

```diff
@@ -1474,3 +1474,158 @@ func TestWalletGETDust(t *testing.T) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		queriedTxns += len(wtg.ConfirmedTransactions)
+	}
+	if queriedTxns != totalTxns {
+		t.Errorf("Expected %v txns but was %v", totalTxns, queriedTxns)
+	}
+}
+
+// TestWalletManyTransactions creates a wallet and sends a large number of
+// coins to itself. Afterwards it will execute subtests to test the wallet's
+// scalability.
+func TestWalletManyTransactions(t *testing.T) {
+	if testing.Short() || !build.VLONG {
+		t.SkipNow()
+	}
+
+	// Declare tests that should be executed
+	subtests := []struct {
+		name string
+		f    func(*testing.T, *serverTester)
+	}{
+		{"TestWalletTransactionEndpoint", testWalletTransactionEndpoint},
+		{"TestWalletTransactionsEndpoint", testWalletTransactionsEndpoint},
+	}
+
+	// Create tester
+	st, err := createServerTester(t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer st.server.panicClose()
+
+	// Disable defrag for the wallet
+	st.wallet.SetSettings(modules.WalletSettings{
+		Defrag: false,
+	})
+
+	// Mining blocks should have created transactions for the wallet containing
+	// miner payouts. Get the list of transactions.
+	var wtg WalletTransactionsGET
+	err = st.getAPI("/wallet/transactions?startheight=0&endheight=-1", &wtg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(wtg.ConfirmedTransactions) == 0 {
+		t.Fatal("expecting a few wallet transactions, corresponding to miner payouts.")
+	}
+	if len(wtg.UnconfirmedTransactions) != 0 {
+		t.Fatal("expecting 0 unconfirmed transactions")
+	}
+
+	// Remember the number of confirmed transactions
+	numConfirmedTxns := len(wtg.ConfirmedTransactions)
+
+	// Get lots of addresses from the wallet
+	numTxns := uint64(10000)
+	ucs, err := st.wallet.NextAddresses(numTxns)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Send SC to each address.
+	minedBlocks := 0
+	for i, uc := range ucs {
+		st.wallet.SendSiacoins(types.SiacoinPrecision, uc.UnlockHash())
+		if i%100 == 0 {
+			if _, err := st.miner.AddBlock(); err != nil {
+				t.Fatal(err)
+			}
+			minedBlocks++
+		}
+	}
+	if _, err := st.miner.AddBlock(); err != nil {
+		t.Fatal(err)
+	}
+	minedBlocks++
+
+	// After sending numTxns times there should be 2*numTxns confirmed
+	// transactions plus one for each mined block. Every send creates a setup
+	// transaction and the actual transaction.
+	err = st.getAPI("/wallet/transactions?startheight=0&endheight=-1", &wtg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	expectedConfirmedTxns := numConfirmedTxns + int(2*numTxns) + minedBlocks
+	if len(wtg.ConfirmedTransactions) != expectedConfirmedTxns {
+		t.Fatalf("expecting %v confirmed transactions but was %v", expectedConfirmedTxns,
+			len(wtg.ConfirmedTransactions))
+	}
+	if len(wtg.UnconfirmedTransactions) != 0 {
+		t.Fatal("expecting 0 unconfirmed transactions")
+	}
+
+	// Execute tests
+	for _, subtest := range subtests {
+		t.Run(subtest.name, func(t *testing.T) {
+			subtest.f(t, st)
+		})
+	}
```

Comment:
I would like to see this test extended so that a giant, blanking reorg is introduced, and then the subtests re-run.

And then another pass where the original chain is mined enough to return to the original chain. Then run the subtests a second time. This will give us confidence that the wallet is correctly handling basic reorgs.

You can set that up by creating a separate node, a blank tester, and mining blocks on it until it has more blocks than this node. When you call `miner.AddBlock()`, the block is returned, which means you can keep the list of blocks without connecting the nodes together. Then you can add all the blocks to this first node via `cs.AcceptBlock()`.

Getting the reorg back to the original chain is a little trickier. You can use the test miner to mine a block without submitting it to the consensus set. You will have to do that, and then you will have to mine another block on top of the first block. Then you submit both blocks to the now-blanked node, and it should reorg back to the original chain, which means all of the transactions should return.